### DISCUSSION TITLE
Redirect not working in mount.

### DISCUSSION THREAD

**Discussion by sohail-panezai** [livewire/livewire#4013] [7. Bugs (archived)] [answered]
### Description
Cannot redirect from mount lifecycle hook.

1. redirect('/') doesn't work.
2. $this->redirect('/') doesn't work.

### Steps to reproduce
1. Make livewire component
2. Redirect from mount method.

**Context**
- Livewire version: 1.2.0
- Laravel version: 5.8.28
- Browser: Chrome

**Comment by nuernbergerA** [comment]
What would be a use case to redirect on mount?

**Comment by acurrieclark** [comment]
I have a case where there is a 3 part form, with the `part` being added as a query string. If the user attempts to go straight to part 3, then I need to redirect them back to 2. 

I can do it in the view, but the URL doesn't update without a redirect.

**Comment by sohail-panezai** [comment]
The component depended upon some data in the session. If that is not present, I redirect. 
I was using
`Route::livewire('route', 'component')`
and changed that to good old controller method in which I checked for that session data and redirected from there.
But, I still hoped it worked.

**Comment by jrleonr** [comment]
Same here! I am using Route::livewire, and I check some values and redirect some times, it kind of stops the execution for mount() but still renders the view.

I had to create a middleware for that but it is annoying...

**Comment by sohail-panezai** [comment]
Caleb had this issue fixed I guess in a previous update. However, its not working on the master branch now.
Check issue [#378 ](378)

**Comment by georgeboot** [comment]
This is actually something we do a lot. Almost all pages with forms (create X page) are livewire components.

If the mount, we usually check some conditions and redirect away if not allowed.

However, livewire will still load the component, but push a redirect instruction in the js payload. In practice, the component loads and some 300ms later, your browser will redirect.

It would be amazing if Livewire would implement "real" redirects here - so that the render method will not even be called.

**Comment by amcsi** [comment]
I'm getting this issue as well.

I simply want to redirect the user to a different page (based on a condition) on landing on my Livewire component/page, but it has no effect in the `mount()` method.

**Comment by Qirel** [comment]
Can confirm that `redirect()` **works** within the `mount()` method in the latest release. If you're not running lower versions for a specific reason and can work around that, I suggest you update your Livewire version. 

Regarding authorization - you can throw exceptions with `abort(403)` / `abort(401)` that will fire before the component is rendered.

**Comment by amcsi** [comment]
@Qirel I can confirm that it doesn't work in the lastest (2.2.5) livewire version.

I call this in `mount()`:

```php
            return redirect()->to(route('guest'));
```

And it just gets ignored, and then the page just gets rendered without redirecting; but the page crashes before actually being able to render, because render depends on the model that gets set up at the end of the happy path of `mount`, which of course when doing redirecting, does not get set.

(And if I put `dd()` above the return redirect code, `dd()` does happen, so I definitely get to the redirect line.)

**Comment by Qirel** [comment]
Hmm.. The test I ran was `return redirect()->route("dasboard");` inside the `mount()` method, and that redirected me just fine. Being on localhost, and that being the only thing inside the method, that worked flawlessly and instantly. 

However, after digging a little more into it and enabling throttling in the network tab, I see that it actually renders the component first, then does the redirect - which is not what I would *expect* it to be, I expected it to run the actual redirect before rendering - so you're correct, it doesn't work prior to the render. Meaning that the redirect is actually performed after the fact..

That said, if you're redirecting away because of insufficient permission,  use the `AuthorizesRequests` trait and call `$this->authorize($ability);`, that would redirect to the login page if not authenticated, or display the 403 page if not authorized. 
Another thing you can do is throwing authorization exceptions through `abort(403)`/`abort(401)`, but I would probably use the first approach.

**Comment by kaloyangdimov** [comment]
Use `$this->redirect('/url');` without returning. Works as expected. Unfortunately I get the redirect after it reloads the page but it might be because of something else. Still  without return you should be fine.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by jadonbrownza** [comment]
I can confirm that using `$this->redirect()` does work, but not as expected.

The first problem is that the page returns a 200 for a redirect. I can't test that the LiveWire component successfully redirects to another page because they're not part of the jetstream team because of this.

The second problem is that the page loads (even the data) before redirecting. In my case, it's sensitive data that's shown from someone else's team and that's a major problem if they're able to screenshot it for the few milliseconds that it shows.

**Comment by rodrigopedra** [comment]
Can confirm @scheMeZa report. Redirects work, but after rendering/returning the payload.

One could have the network tab open and just look at the rendered response from the request. 

From my understanding the redirect is done after render by JavaScript when redirecting from `mount()`.

I tried throwing a `HttpResponseException`, but that also didn't work.

~~~php
use Illuminate\Http\Exceptions\HttpResponseException;
use Illuminate\Http\RedirectResponse;

throw new HttpResponseException(new RedirectResponse(\route('dashboard')));
~~~

If there is a better way to redirect before even returning any response, please let me know. Thanks!

**Comment by rodrigopedra** [comment]
REOPEN

**Comment by rodrigopedra** [comment]
As a workaround I got this working:

1 - Create a `blank` blade view file with just an empty `<div>` element

~~~
<div></div>
~~~

2 - Check if `$this->redirectTo` is set, if so render that blank view

~~~php
<?php

namespace App\Http\Livewire;

use App\Models\User;
use Livewire\Component;

class MyComponent extends Component
{
    public $records;

    public function mount()
    {
        if (/* condition is met */) {
            return redirect()->route('login');
        }

        // do any initialization AFTER checking for redirection
        $this->records = User::all();
    }

    public function render()
    {
        if ($this->redirectTo) {
            // if you need to do any initialization before redirecting,
            // undo them before returning the view, otherwise data will
            // be leaked to the component's wire:initial-data attribute
            // NOT NEEDED if you initialized after returning the redirect on mount()
            $this->records = null;
            
            return \view('blank');
        }

        return view('livewire.my-component');
    }
}
~~~

Please consider the comments in the snippet above to avoid any data leaking.

**Comment by calebporzio** [comment]
Should be fixed in the next release see #2223

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by rodrigopedra** [comment]
Thanks @calebporzio and @nuernbergerA

**Comment by hazem-taha** [comment]
I'm still not getting this to work, this is so weird, how this issue is still not fixed yet :(

**Comment by kenlas** [comment]
REOPEN

**Comment by kenlas** [comment]
I think there is an edge case, when a redirect is issued in mount() and handled well by the Component but I think in this case the render() function should not be called at all.

This is how it is handled now:
```
$manager = LifecycleManager::fromInitialInstance($this)
    ->initialHydrate()
    ->mount($componentParams)
    ->renderToView();

if ($this->redirectTo) {
    return redirect()->response($this->redirectTo);
}
```
But it's calling `renderToView()` anyway.

I would prefer calling `renderToView()` after checking `$this->redirectTo`. If I skip mounting logic with a redirect after checking some conditions, calling the render function with the un-initialized variables, etc is unnecessary.

@calebporzio what is your opinion? Should I make a pull request for this?

**Reply by joshhanley** [reply]
This issue has been converted to a discussion as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/livewire/livewire/discussions/3147

**Comment by joshhanley** [comment, marked as answer]
@kenlas skip render on redirect was removed in the last release, which impacts the redirect in mount.

So this functionality has been restored and will be in the next release 🙂
See PR #3971 for details.

Hope this helps!

**Reply by kenlas** [reply]
Thanks @joshhanley this is exactly what I was looking for!

**Reply by joshhanley** [reply]
No worries 🙂

**Comment by michael-rubel** [comment]
Got an `Undefined array key 1` error after redirecting inside the `mount` method from the `nested` component of a `full-page` component.

Tried:
- Classical `redirect()->to()`
- `$this->redirect()`
- `$this->redirectRoute()`

PHP version: `8.1`
Laravel version: `v9.3.1`
Livewire version: `v2.10.4`

File/line: RendersLivewireComponents.php: 83

![scr](https://user-images.githubusercontent.com/37669560/156755092-466f9c3f-4482-419d-a4bf-8b50401dfb1d.png)

**Comment by snapey** [comment]
Still having this problem.  My component requires a valid model id to be passed into the mount method, but the redirect never occurs (when user tampers with the url)
```
    public function mount()
    {
        try {
            $this->new_title = Defect::where('uuid',$this->uuid)->firstOrFail()->title;
        } catch (\Throwable $th) {
            $this->skipRender();
            $this->redirect(route('defects.index'));
        }
    }
```

The result  of passing in a missing model is that the page is rendered minus the livewire component view, and never redirects.

Livewire 2.10.5
Laravel 9.7.0
PHP 8.1.3

**Reply by joshhanley** [reply]
@snapey I'm not sure I follow exactly, is this in a nested component? any chance you could share a small example repo or submit a failing test so we can investigate it?

Thanks!

**Reply by snapey** [reply]
No, it's not nested.

I'm expecting a model UUID to be passed into the livewire component from the route.  If that UUID is tampered with and does not represent a model then I want to bail and redirect somewhere else.  The example in the mount() method results in empty component content and no redirect

**Reply by AmirHoseinBahrami** [reply]
thanks

**Reply by wilsonseixo** [reply]
Any news? Were you able to solve it?

**Reply by Spryer** [reply]
welp nothing new?

**Reply by lrljoe** [reply]
@Spryer <- what's the use case here?   You want to mount a component, then redirect them to another route on certain circumstances?  You've asked "nothing new?", but are asking on
a) An older thread
b) Not actually articulating the issue you're facing exactly.

@joshhanley shared a **very** precise example of how to do this:
```
public function mount() {
    $this->skipRender();
    return $this->redirect('/some-path');
}
```

If you are using nested components, and needing to do that, then it's indicative of flaws in your application logic, which can be addressed further up the chain.  Loading a component in, just to redirect somewhere else, seems a bit of a waste!

If you want a multi-stage form, then Spatie have a package precisely for that.

**Reply by Spryer** [reply]
my bad. I already solve the redirection error with using a Controller instead of a Livewire Component :)
